internal/core/software_renderer.rs

```diff
@@ -1129,20 +1178,84 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
     }
 
     fn process_texture_impl(&mut self, geometry: PhysicalRect, texture: SceneTexture<'_>) {
-        self.foreach_ranges(&geometry, |line, buffer, extra_left_clip, extra_right_clip| {
-            draw_functions::draw_texture_line(
-                &geometry,
-                PhysicalLength::new(line),
-                &texture,
-                buffer,
-                extra_left_clip,
-                extra_right_clip,
-            );
+        self.foreach_region(&geometry, |buffer, rect, extra_left_clip, extra_right_clip| {
+            if !buffer.draw_texture(
+                rect.origin.x,
+                rect.origin.y,
+                rect.size.width,
+                rect.size.height,
+                geometry.origin.y,
+                target_pixel_buffer::Texture {
+                    bytes: texture.data,
+                    pixel_format: texture.format,
+                    pixel_stride: texture.pixel_stride,
+                    width: texture.source_size().width as u16,
+                    height: texture.source_size().height as u16,
+                    delta_x: texture.extra.dx.0,
+                    delta_y: texture.extra.dy.0,
+                    source_offset_x: texture.extra.off_x.0,
+                    source_offset_y: texture.extra.off_y.0,
+                },
+                texture.extra.colorize.as_argb_encoded(),
+                texture.extra.alpha,
+                texture.extra.rotation,
+                CompositionMode::default(),
+            ) {
+                let begin = rect.min_x();
+                let end = rect.max_x();
+                for l in rect.min_y()..rect.max_y() {
+                    draw_functions::draw_texture_line(
+                        &geometry,
+                        PhysicalLength::new(l),
+                        &texture,
+                        &mut buffer.line_slice(l as usize)[begin as usize..end as usize],
+                        extra_left_clip,
+                        extra_right_clip,
+                    );
+                }
+            }
         });
     }
+
+    fn process_rectangle_impl(
+        &mut self,
+        geometry: PhysicalRect,
+        color: PremultipliedRgbaColor,
+        composition_mode: CompositionMode,
+    ) {
+        if !self.buffer.fill_rectangle(
+            geometry.origin.x,
+            geometry.origin.y,
+            geometry.size.width,
+            geometry.size.height,
+            color,
+            composition_mode,
+        ) {
+            self.foreach_region(&geometry, |buffer, rect, _extra_left_clip, _extra_right_clip| {
+                let begin = rect.min_x();
+                let end = rect.max_x();
+                for l in rect.min_y()..rect.max_y() {
+                    match composition_mode {
+                        CompositionMode::Source => {
+                            let mut fill_col = B::TargetPixel::background();
+                            B::TargetPixel::blend(&mut fill_col, color);
```

Comment:
It'd be faster to put that outside of the loop.  (meaning also swapping the `match` and the `for`

Reply:
I was contemplating that and thinking the compiler will do that automatically. But swapping match and for is a good idea anyway.